src/core/services/interimElement/interimElementProvider.js

```diff
@@ -0,0 +1,274 @@
+    }
+  }
+}
+
+function InterimElementService($$interimElementBase, $q) {
+  // This is needed, otherwise ng-annotator can't detect this function as a factory $get.
+  'ngInject';
+
+  return function() {
+    var API;
+    var SHOW_CANCELLED = false;
+    var showQueue = [];
+
+    return API = {
+      showInterim: _showInterim,
+      hideInterim: _hideInterim,
+      cancelInterim: _cancelInterim,
+      destroyInterim: _destroyInterim
+    };
+
+    function _showInterim(options) {
+      options = options || {};
+      // We inject the actual interim service into the options.
+      // This is needed for hideDelay etc. because it will call the service to hide the interim.
+      // This important, because otherwise the showQueue won't be cleared.
+      options._service = API;
+
+      var interimElement = new $$interimElementBase(options || {});
+      var hideExisting = !options.skipHide && showQueue.length ? API.hideInterim() : $q.when(true);
+
+      // Wait for the current existing interim to be hidden.
+      hideExisting.finally(function() {
+
+        showQueue.push(interimElement);
+
+        interimElement.show();
+
+      });
+
+      // Return a promise that will be resolved when the interim
+      // element is hidden or cancelled.
+      return interimElement.deferred.promise;
+    }
+
+    function _hideInterim(reason, options) {
+      if (!showQueue.length) return $q.when(reason);
+      options = options || {};
+
+      if (options.closeAll) {
+        var promise = $q.all(showQueue.reverse().map(closeElement));
+        showQueue = [];
+        return promise;
+      } else if (options.closeTo !== undefined) {
+        return $q.all(showQueue.splice(options.closeTo).map(closeElement));
+      }
+
+      var interim = showQueue.pop();
+      return closeElement(interim);
+
+      function closeElement(interim) {
+
+        interim.remove(reason, false, options);
+
+        return interim.deferred.promise;
+      }
+    }
+
+    function _cancelInterim(reason, options) {
+      var interim = showQueue.pop();
+      if (!interim) return $q.when(reason);
+
+      interim.remove(reason, true, options || {});
+
+      return interim.deferred.promise;
+    }
+
+    function _destroyInterim(target) {
+      var interim = !target ? showQueue.shift() : null;
+
+      var cntr = angular.element(target).length ? target.parentNode : null;
+
+      if (cntr) {
+
+        // Try to find the interim element in the stack which corresponds to the supplied DOM element.
+        var filtered = showQueue.filter(function(entry) {
+          var currNode = entry.options.element[0];
+          return currNode === cntr;
+        });
+
+        // Note: this function might be called when the element already has been removed, in which
+        // case we won't find any matches. That's ok.
+        if (filtered.length > 0) {
+          interim = filtered[0];
+          showQueue.splice(showQueue.indexOf(interim), 1);
```

Comment:
It could be a good idea to checkout if the index here is bigger than -1. If the `interim` isn't in the `showQueue`, the splice will still execute and return an array with only the last element.